<a href="https://colab.research.google.com/github/avagup/customer-master-data-management/blob/main/Customer_data_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')
folder_path = '/content/drive/MyDrive/Colab Notebooks/Personal Projects/'


Mounted at /content/drive


In [2]:
df = pd.read_csv(folder_path+'customer_data_raw.csv', low_memory=False)
df.info()
df.shape
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10180 entries, 0 to 10179
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Customer_ID      10180 non-null  object
 1   First_Name       10100 non-null  object
 2   Last_Name        10180 non-null  object
 3   Email            10061 non-null  object
 4   Phone            9924 non-null   object
 5   Date_of_Birth    10025 non-null  object
 6   Country          10180 non-null  object
 7   State            10078 non-null  object
 8   Signup_Date      10180 non-null  object
 9   Customer_Status  10180 non-null  object
dtypes: object(10)
memory usage: 795.4+ KB


,Customer_ID,First_Name,Last_Name,Email,Phone,Date_of_Birth,Country,State,Signup_Date,Customer_Status
0,CUST008508,Amelia,Smith,AMELIA.SMITH9508@EXAMPLE.COM,+1-4769407963,1990-05-28,USA,Illinois,2020-09-15,Active
1,CUST003131,William,Roberts,william.roberts4131@example.com,+1-6088963087,1991-12-19,Canada,Ontario,2025-03-29,Active
2,CUST000719,Ibrahim,Gupta,ibrahim.gupta1719@example.com,+1-5977144010,2004-06-11,USA,New York,2024-05-19,Inactive
3,CUST009850,Olivia,Harris,olivia.harris10850@example.com,+1-2477020276,1958-07-06,Canada,Quebec,2021-07-25,Cancelled
4,CUST002761,Isabella,Singh,isabella.singh3761@example.com,+65-97133082,1964-06-15,Singapore,North Region,2025-02-04,Canceled


In [3]:
df.describe(include='all')

,Customer_ID,First_Name,Last_Name,Email,Phone,Date_of_Birth,Country,State,Signup_Date,Customer_Status
count,10180,10100,10180,10061,9924,10025,10180,10078,10180,10180
unique,10100,106,80,9773,9750,7628,31,63,2332,12
top,CUST009489,Priya,Mehta,not_an_email,+91-7403390846,2028-01-15,India,Delhi,2027-03-10,Active
freq,2,318,425,28,2,25,2812,489,21,6101


In [4]:
(df.isnull().sum() / len(df) * 100).round(2)

,0
Customer_ID,0.00
First_Name,0.79
Last_Name,0.00
Email,1.17
Phone,2.51
Date_of_Birth,1.52
Country,0.00
State,1.00
Signup_Date,0.00
Customer_Status,0.00


In [5]:
print(f"Unique CustIDs = {df["Customer_ID"].nunique()}")
print(f"Unique Emails = {df["Email"].nunique()}")

print(df.duplicated().sum())

Unique CustIDs = 10100
Unique Emails = 9773
75


In [6]:
#Check no. of missing values
df.isnull().sum()

,0
Customer_ID,0
First_Name,80
Last_Name,0
Email,119
Phone,256
Date_of_Birth,155
Country,0
State,102
Signup_Date,0
Customer_Status,0


In [7]:
print(df["Country"].value_counts(dropna=False))
print(df["Customer_Status"].value_counts(dropna=False))
print(df["State"].value_counts(dropna=False))

Country
India             2812
USA               2508
UK                1200
Canada            1025
Australia          790
Germany            786
Singapore          734
  India             31
INDIA               27
usa                 25
india               25
  USA               23
canada              18
Ind                 18
United Kingdom      15
  Germany           13
U.S.A.              13
  UK                12
GERMANY             12
uk                  12
singapore           11
germany             11
AUSTRALIA           10
SINGAPORE           10
CANADA              10
australia            8
  Canada             7
  Singapore          6
  Australia          6
  india              1
  usa                1
Name: count, dtype: int64
Customer_Status
Active       6101
Inactive     2050
Cancelled    1254
Suspended     592
active         56
ACTIVE         50
INACTIVE       24
inactive       21
Canceled       11
cancelled      10
SUSPENDED       7
suspended       4
Name: count, dtype: i

In [8]:
#A. Date Conversions
df["Date_of_Birth"] = pd.to_datetime(df["Date_of_Birth"], errors="coerce")
df["Signup_Date"] = pd.to_datetime(df["Signup_Date"], errors="coerce")

print(df["Date_of_Birth"].min())
print(df["Date_of_Birth"].max())

print(df["Signup_Date"].min())
print(df["Signup_Date"].max())


print(df[df["Date_of_Birth"] > pd.Timestamp.today()])
print(df[df["Signup_Date"] > pd.Timestamp.today()])

#Suspicious dates - 2028-01-15 DOB, 2027-03-10 Signup date

1955-01-01 00:00:00
2028-01-15 00:00:00
2019-01-01 00:00:00
2027-03-10 00:00:00
     Customer_ID First_Name Last_Name                            Email  \
228   CUST002158     Amelia     Davis     amelia.davis3158@example.com   
451   CUST004984       Noah      Iyer        noah.iyer5984@example.com   
486   CUST000857      Rahul  Williams   rahul.williams1857@example.com   
790   CUST000458       Liam     Kumar       liam.kumar1458@example.com   
822   CUST005690       Emma     Brown       emma.brown6690@example.com   
1051  CUST006623      Karan     Smith      karan.smith7623@example.com   
1106  CUST006253      Sarah    Walker     sarah.walker7253@example.com   
2066  CUST007357      Ethan      Shah       ethan.shah8357@example.com   
2180  CUST004449      Grace    Taylor     grace.taylor5449@example.com   
2394  CUST000046     amelia  Williams  amelia.williams1046@example.com   
2495  CUST002956       john    Wilson      john.wilson3956@example.com   
2730  CUST004780   Isabella    T

In [9]:
#B. Email
email_pattern = r"^[^@\s]+@[^@\s]+\.[^@\s]+$"

invalid_email = df[
    df["Email"].notna() &
    ~df["Email"].str.strip().str.match(email_pattern , na=False)
]

print(len(invalid_email))
print(f"invalid formats - {invalid_email["Email"].value_counts()}")

104
invalid formats - Email
user@                    28
not_an_email             28
user.example.com         24
missing_at_symbol.com    22
  user@                   1
  user.example.com        1
Name: count, dtype: int64


In [10]:
df["Email_Normalized"] = (
    df["Email"]
    .astype("string")
    .str.strip()
    .str.lower()
)

print(f"No. of dup emails = {df['Email_Normalized'].duplicated().sum()}")

duplicate_emails = df[
    df["Email_Normalized"].duplicated(keep=False)
].sort_values("Email_Normalized")

duplicate_emails.head(20)

No. of dup emails = 412


,Customer_ID,First_Name,Last_Name,Email,Phone,Date_of_Birth,Country,State,Signup_Date,Customer_Status,Email_Normalized
5568,CUST010063,Aarav,Gupta,aarav.gupta4976@example.com,+91-8534142254,2000-07-24,India,Gujarat,2024-06-16,Active,aarav.gupta4976@example.com
6262,CUST003976,Aarav,Gupta,aarav.gupta4976@example.com,+91-8534142254,2000-07-24,India,Gujarat,2024-06-16,Active,aarav.gupta4976@example.com
112,CUST010044,Aarav,Harris,aarav.harris10768@example.com,+1-7865848061,1976-03-12,USA,Washington,2024-04-14,Active,aarav.harris10768@example.com
5174,CUST009768,Aarav,Harris,aarav.harris10768@example.com,+1-7865848061,1976-03-12,USA,Washington,2024-04-14,Active,aarav.harris10768@example.com
9862,CUST010083,Aarav,Johnson,aarav.johnson5740@example.com,+65-87866323,1991-07-12,Singapore,Central Region,2020-12-11,Active,aarav.johnson5740@example.com
5917,CUST004740,Aarav,Johnson,aarav.johnson5740@example.com,+65-87866323,1991-07-12,Singapore,Central Region,2020-12-11,Active,aarav.johnson5740@example.com
1729,CUST010042,Aarav,Kumar,aarav.kumar4269@example.com,+61-490490785,1972-02-05,Australia,Queensland,2025-06-14,Active,aarav.kumar4269@example.com
4981,CUST003269,Aarav,Kumar,aarav.kumar4269@example.com,+61-490490785,1972-02-05,Australia,Queensland,2025-06-14,Active,aarav.kumar4269@example.com
9847,CUST003010,Aarav,Thomas,aarav.thomas4010@example.com,+49-1517916477,1995-06-10,Germany,Hamburg,2024-12-01,Suspended,aarav.thomas4010@example.com
2224,CUST010035,Aarav,Thomas,aarav.thomas4010@example.com,+49-1517916477,1995-06-10,Germany,Hamburg,2024-12-01,Suspended,aarav.thomas4010@example.com


In [11]:
#For each duplicated email, how many different Customer IDs are associated with it?
duplicate_emails.groupby("Email_Normalized")["Customer_ID"].nunique().value_counts()

,count
Customer_ID,
2,98
1,79
18,1
22,1
28,1
25,1
29,1


In [12]:
#This will show us cases where the same normalized email belongs to different Customer IDs.
duplicate_emails[
    duplicate_emails.groupby("Email_Normalized")["Customer_ID"]
    .transform("nunique") > 1
].head(20)

#Duplicate customer records were identified through normalized email addresses.
#Multiple Customer IDs share the same normalized email, and several of these records have identical
#customer attributes, indicating duplicate customer records rather than legitimate shared-email cases.

,Customer_ID,First_Name,Last_Name,Email,Phone,Date_of_Birth,Country,State,Signup_Date,Customer_Status,Email_Normalized
5568,CUST010063,Aarav,Gupta,aarav.gupta4976@example.com,+91-8534142254,2000-07-24,India,Gujarat,2024-06-16,Active,aarav.gupta4976@example.com
6262,CUST003976,Aarav,Gupta,aarav.gupta4976@example.com,+91-8534142254,2000-07-24,India,Gujarat,2024-06-16,Active,aarav.gupta4976@example.com
112,CUST010044,Aarav,Harris,aarav.harris10768@example.com,+1-7865848061,1976-03-12,USA,Washington,2024-04-14,Active,aarav.harris10768@example.com
5174,CUST009768,Aarav,Harris,aarav.harris10768@example.com,+1-7865848061,1976-03-12,USA,Washington,2024-04-14,Active,aarav.harris10768@example.com
9862,CUST010083,Aarav,Johnson,aarav.johnson5740@example.com,+65-87866323,1991-07-12,Singapore,Central Region,2020-12-11,Active,aarav.johnson5740@example.com
5917,CUST004740,Aarav,Johnson,aarav.johnson5740@example.com,+65-87866323,1991-07-12,Singapore,Central Region,2020-12-11,Active,aarav.johnson5740@example.com
1729,CUST010042,Aarav,Kumar,aarav.kumar4269@example.com,+61-490490785,1972-02-05,Australia,Queensland,2025-06-14,Active,aarav.kumar4269@example.com
4981,CUST003269,Aarav,Kumar,aarav.kumar4269@example.com,+61-490490785,1972-02-05,Australia,Queensland,2025-06-14,Active,aarav.kumar4269@example.com
9847,CUST003010,Aarav,Thomas,aarav.thomas4010@example.com,+49-1517916477,1995-06-10,Germany,Hamburg,2024-12-01,Suspended,aarav.thomas4010@example.com
2224,CUST010035,Aarav,Thomas,aarav.thomas4010@example.com,+49-1517916477,1995-06-10,Germany,Hamburg,2024-12-01,Suspended,aarav.thomas4010@example.com


In [13]:
#C. Phone Number
df["Phone_Digits"] = (
    df["Phone"]
    .astype("string")
    .str.replace(r"\D", "", regex=True)
)

#df["Phone_Digits"].str.len().value_counts(dropna=False)

df.groupby("Country")["Phone_Digits"].apply(
    lambda x: x.str.len().value_counts(dropna=False)
)

Country             
  Australia     11         5
                <NA>       1
  Canada        11         7
  Germany       12        13
  India         12        29
                <NA>       2
  Singapore     10         6
  UK            12        12
  USA           11        21
                <NA>       2
  india         12         1
  usa           11         1
AUSTRALIA       11        10
Australia       11       767
                <NA>      23
CANADA          11        10
Canada          11       998
                <NA>      27
GERMANY         12        12
Germany         12       766
                <NA>      20
INDIA           12        25
                <NA>       2
Ind             12        18
India           12      2747
                <NA>      65
SINGAPORE       10        10
Singapore       10       717
                <NA>      17
U.S.A.          11        13
UK              12      1167
                <NA>      33
USA             11      2450
                <NA>      58
United Kingdom  12        15
australia       11         7
                <NA>       1
canada          11        18
germany         12         9
                <NA>       2
india           12        24
                <NA>       1
singapore       10        11
uk              12        11
                <NA>       1
usa             11        24
                <NA>       1
Name: Phone_Digits, dtype: Int64

In [14]:
print(df[["Country", "Phone"]].dropna().sample(30, random_state=42))

#Phone formatting is inconsistent. Phone numbers are represented using different formats,
#including +country-code-number, digits-only values, and space-separated values.
#This inconsistency can make matching, deduplication, validation, and downstream analysis more difficult.

        Country           Phone
5410        USA   +1-4491820146
528         USA   +1-4769226185
4445  Australia   +61-416220530
8601         UK  +44-7069857956
9772        USA   +1-8491120501
8870      India  +91-7469024143
8566     Canada   +1-8473805200
4080     Canada   +1-9252844754
9424     Canada   142 436 87092
1347        USA   +1-7754754346
8080         UK    447782121977
6565  Australia   +61-438527425
7448      India  +91-9379761013
2716        USA   +1-7567535997
5263         UK  447 598 928086
3526        USA   +1-6335918133
8824      India  +91-8376672784
5296    Germany  +49-1729883642
9964  Singapore    +65-92437906
2302    Germany  +49-1724671822
816       India  +91-6247790526
9336        USA   +1-4825454480
8964        USA   +1-4523840012
1208        USA   137 087 72809
3784         UK  +44-7246786920
1858  Australia   +61-462564019
36        India  +91-6769626461
4202        USA   +1-3087180863
7266      India  +91-6448442769
3080      India  +91-9096513077


In [15]:
#D. Customer duplicates
duplicate_ids = (
    df[df["Customer_ID"].duplicated(keep=False)]
    .sort_values("Customer_ID")
)

duplicate_ids.head()

,Customer_ID,First_Name,Last_Name,Email,Phone,Date_of_Birth,Country,State,Signup_Date,Customer_Status,Email_Normalized,Phone_Digits
3797,CUST000206,Daniel,Davis,daniel.davis1206@example.com,447444756508,1971-01-06,UK,England,2021-01-26,Inactive,daniel.davis1206@example.com,447444756508
6189,CUST000206,Daniel,Davis,daniel.davis1206@example.com,447444756508,1971-01-06,UK,England,2021-01-26,Inactive,daniel.davis1206@example.com,447444756508
8650,CUST000208,Olivia,Williams,olivia.williams1208@example.com,+91-7922352161,1971-08-01,India,Karnataka,2022-07-13,Active,olivia.williams1208@example.com,917922352161
346,CUST000208,Olivia,Williams,olivia.williams1208@example.com,+91-7922352161,1971-08-01,India,Karnataka,2022-07-13,Active,olivia.williams1208@example.com,917922352161
2858,CUST000399,Grace,Kumar,grace.kumar1399@example.com,+91-9203563846,1962-06-13,India,Telangana,2024-08-08,Inactive,grace.kumar1399@example.com,919203563846


In [16]:
duplicate_ids["Customer_ID"].value_counts()

#Duplicate Customer_IDs are present. Some duplicated IDs represent exact duplicate records, while others contain redundant records with minor differences.
#Duplicate records should be investigated and resolved before downstream analysis.

,count
Customer_ID,
CUST000206,2
CUST000208,2
CUST000399,2
CUST000464,2
CUST000497,2
...,...
CUST009227,2
CUST009381,2
CUST009443,2


In [17]:
#E. Customer Status

df["Customer_Status"].value_counts(dropna=False)

# Customer_Status contains inconsistent representations of the same business statuses. Variations include differences in capitalization (Active, active, ACTIVE) and spelling (Cancelled, Canceled).
#These should be standardized to a defined set of canonical status values.

,count
Customer_Status,
Active,6101
Inactive,2050
Cancelled,1254
Suspended,592
active,56
ACTIVE,50
INACTIVE,24
inactive,21
Canceled,11


In [18]:
#########CORRECTIONS############

In [19]:
# INVALID EMAILs
#Invalid email format: Flag the email as invalid and retain the customer record.
#Do not automatically infer or fabricate a replacement email.
#If an authoritative source is available, use it to validate or update the value.



In [20]:
# future DOB and future Signup_Date
#Flag the record as invalid. If the field is permitted to be null, set the invalid value to missing.
#Attempt correction only when the correct value can be reliably established from an authoritative source or documented business rule.
#Otherwise, retain as missing and flag for review.

In [21]:
#Phone formatting
#Standardize phone numbers by removing non-digit characters and unnecessary spaces while preserving the underlying number.
#Do not infer, fabricate, or alter missing digits.

In [22]:
#IMPLEMENTATION

In [23]:
df["Country"] = df["Country"].str.strip()

country_mapping = {
    "India": "India",
    "india": "India",
    "INDIA": "India",
    "Ind": "India",
    "  india":"India",

    "  USA": "USA",
    "usa": "USA",
    "U.S.A.": "USA",

    "UK": "UK",
    "uk": "UK",
    "United Kingdom": "UK",

    "Australia": "Australia",
    "australia": "Australia",
    "AUSTRALIA": "Australia",
    "  Australia":"Australia",

    "Canada": "Canada",
    "canada": "Canada",
    "  Canada":"Canada",
    "CANADA": "Canada",

    "Germany": "Germany",
    "germany": "Germany",
    "GERMANY": "Germany",
    "  Germany": "Germany",

    "Singapore": "Singapore",
    "singapore": "Singapore",
    "SINGAPORE": "Singapore",
    "  Singapore": "Singapore",

}

df["Country"] = df["Country"].replace(country_mapping)

df["Country"].value_counts(dropna=False)

,count
Country,
India,2914
USA,2570
UK,1239
Canada,1060
Germany,822
Australia,814
Singapore,761


In [24]:
df.shape

(10180, 12)

In [25]:
df["Customer_Status"] = df["Customer_Status"].str.strip()

status_mapping = {
    "Active": "Active",
    "active": "Active",
    "ACTIVE": "Active",
    "INACTIVE": "Inactive",
    "inactive": "Inactive",
    "Cancelled": "Cancelled",
    "canceled": "Cancelled",
    "Canceled": "Cancelled",
    "cancelled": "Cancelled",
    "CANCELED": "Cancelled",
    "SUSPENDED": "Suspended",
    "suspended": "Suspended"

}

df["Customer_Status"] = df["Customer_Status"].replace(status_mapping)

df["Customer_Status"].value_counts(dropna=False)

,count
Customer_Status,
Active,6207
Inactive,2095
Cancelled,1275
Suspended,603


In [26]:
#Exact duplicate records → keep one occurrence and remove subsequent identical copies.
df.duplicated().sum()

np.int64(76)

In [27]:
df = df.drop_duplicates(keep="first")

In [28]:
print("Rows after removing exact duplicates:", len(df))
print("Remaining exact duplicates:", df.duplicated().sum())

Rows after removing exact duplicates: 10104
Remaining exact duplicates: 0


In [29]:
duplicate_ids = (
    df[df["Customer_ID"].duplicated(keep=False)]
    .sort_values("Customer_ID")
)

duplicate_ids["Customer_ID"].value_counts()

,count
Customer_ID,
CUST000206,2
CUST002751,2
CUST004572,2
CUST007527,2


In [30]:
#We want to inspect whether the remaining duplicates have:
#exactly the same customer information,
#missing values in one record but populated values in another,
#or genuinely conflicting information such as different phone numbers, emails, DOBs, etc.

duplicate_ids.head(20)

,Customer_ID,First_Name,Last_Name,Email,Phone,Date_of_Birth,Country,State,Signup_Date,Customer_Status,Email_Normalized,Phone_Digits
3797,CUST000206,Daniel,Davis,daniel.davis1206@example.com,447444756508,1971-01-06,UK,England,2021-01-26,Inactive,daniel.davis1206@example.com,447444756508
6189,CUST000206,Daniel,Davis,daniel.davis1206@example.com,447444756508,1971-01-06,UK,England,2021-01-26,Inactive,daniel.davis1206@example.com,447444756508
2696,CUST002751,Sarah,Patel,sarah.patel3751@example.com,+1-5003104854,1990-04-08,USA,Washington,2020-04-17,Inactive,sarah.patel3751@example.com,15003104854
8078,CUST002751,Sarah,Patel,sarah.patel3751@example.com,+1-5003104854,1990-04-08,USA,Washington,2020-04-17,Inactive,sarah.patel3751@example.com,15003104854
1270,CUST004572,Riya,Wilson,riya.wilson5572@example.com,+61-403699940,1975-10-31,Australia,New South Wales,2022-09-29,Active,riya.wilson5572@example.com,61403699940
7487,CUST004572,Riya,Wilson,riya.wilson5572@example.com,+61-403699940,1975-10-31,Australia,New South Wales,2022-09-29,Active,riya.wilson5572@example.com,61403699940
6190,CUST007527,Ethan,Garcia,ethan.garcia8527@example.com,+91-8235878656,1970-11-01,India,Maharashtra,2021-04-21,Cancelled,ethan.garcia8527@example.com,918235878656
10021,CUST007527,Ethan,Garcia,ethan.garcia8527@example.com,+91-8235878656,1970-11-01,India,Maharashtra,2021-04-21,Cancelled,ethan.garcia8527@example.com,918235878656


In [31]:
text_cols = [
    "First_Name",
    "Last_Name",
    "Email",
    "Country",
    "State",
    "Customer_Status"
]

for col in text_cols:
    df[col] = df[col].str.strip()

df.duplicated().sum()

np.int64(4)

In [32]:
df = df.drop_duplicates(keep="first")
print("Rows:", len(df))
print("Exact duplicates:", df.duplicated().sum())

Rows: 10100
Exact duplicates: 0


In [33]:
duplicate_ids = df[
    df["Customer_ID"].duplicated(keep=False)
].sort_values("Customer_ID")

print("Rows involved in duplicate Customer_IDs:", len(duplicate_ids))
print("Duplicate Customer_IDs:", duplicate_ids["Customer_ID"].nunique())

Rows involved in duplicate Customer_IDs: 0
Duplicate Customer_IDs: 0


In [34]:
# Phone Number

df["Phone_Digits"] = (
    df["Phone"]
    .astype("string")
    .str.replace(r"\D", "", regex=True)
)

df[["Phone", "Phone_Digits"]].head(20)

,Phone,Phone_Digits
0,+1-4769407963,14769407963
1,+1-6088963087,16088963087
2,+1-5977144010,15977144010
3,+1-2477020276,12477020276
4,+65-97133082,6597133082
5,+65-94873761,6594873761
6,+1-7566976541,17566976541
7,+44-7140459628,447140459628
8,917672680066,917672680066
9,+91-7753415268,917753415268


In [35]:
df["Phone_Digits"].isna().sum()

np.int64(255)

In [36]:
df["Phone_Digits"].str.len().value_counts(dropna=False).sort_index()

df.groupby("Country")["Phone_Digits"].apply(
    lambda x: x.str.len().value_counts(dropna=False)
)

#Phone numbers were normalized to digits-only format and validated against country-specific expected lengths.
#All non-null normalized phone numbers in the current dataset conform to the expected length
#for their associated country.

Country        
Australia  11       782
           <NA>      25
Canada     11      1025
           <NA>      27
Germany    12       797
           <NA>      22
India      12      2819
           <NA>      70
Singapore  10       733
           <NA>      17
UK         12      1196
           <NA>      34
USA        11      2493
           <NA>      60
Name: Phone_Digits, dtype: Int64

In [37]:
#EMAIL

df["Email_Normalized"].isna().sum()


np.int64(118)

In [38]:
df["Email_Normalized"].str.contains(
    r"^[^@\s]+@[^@\s]+\.[^@\s]+$",
    regex=True,
    na=False
).value_counts()

,count
Email_Normalized,
True,9878
False,222


In [39]:
email_pattern = r"^[^@\s.]+(\.[^@\s.]+)*@[^@\s.]+(\.[^@\s.]+)+$"

df["Email_Valid"] = df["Email_Normalized"].str.match(
    email_pattern,
    na=False
)

df["Email_Valid"].value_counts()


,count
Email_Valid,
True,9860
False,240


In [40]:
df["Email_Quality"] = "Valid"

df.loc[df["Email_Normalized"].isna(), "Email_Quality"] = "Missing"

df.loc[
    df["Email_Normalized"].notna() & ~df["Email_Valid"],
    "Email_Quality"
] = "Invalid"

In [41]:
df["Email_Quality"].value_counts()

,count
Email_Quality,
Valid,9860
Invalid,122
Missing,118


In [42]:
# DATE OF BIRTH

df["Date_of_Birth"].isna().sum()

np.int64(154)

In [43]:
df["Date_of_Birth"].gt(pd.Timestamp.today()).sum()

np.int64(25)

In [44]:
df.loc[
    df["Date_of_Birth"] > pd.Timestamp.today(),
    ["Customer_ID", "Date_of_Birth", "Signup_Date", "Country"]
]

,Customer_ID,Date_of_Birth,Signup_Date,Country
228,CUST002158,2028-01-15,2024-05-15,USA
451,CUST004984,2028-01-15,2020-04-28,USA
486,CUST000857,2028-01-15,2022-02-18,Australia
790,CUST000458,2028-01-15,2023-10-27,UK
822,CUST005690,2028-01-15,2024-11-30,USA
1051,CUST006623,2028-01-15,2022-06-10,India
1106,CUST006253,2028-01-15,2021-02-28,India
2066,CUST007357,2028-01-15,2020-05-17,UK
2180,CUST004449,2028-01-15,2019-01-16,Singapore
2394,CUST000046,2028-01-15,2022-09-11,India


In [45]:
future_dob = df["Date_of_Birth"] > pd.Timestamp.today()

df["DOB_Quality"] = "Valid"

df.loc[df["Date_of_Birth"].isna(), "DOB_Quality"] = "Missing"
df.loc[future_dob, "DOB_Quality"] = "Invalid_Future"

df.loc[future_dob, "Date_of_Birth"] = pd.NaT

In [46]:
df["DOB_Quality"].value_counts()

,count
DOB_Quality,
Valid,9921
Missing,154
Invalid_Future,25


In [47]:
# SIGNUP DATE

df["Signup_Date"].isna().sum()

np.int64(0)

In [48]:
future_signup = df["Signup_Date"] > pd.Timestamp.today()
future_signup.sum()

np.int64(21)

In [49]:
df.loc[
    future_signup,
    ["Customer_ID", "Date_of_Birth", "Signup_Date", "Country", "Customer_Status"]
]

,Customer_ID,Date_of_Birth,Signup_Date,Country,Customer_Status
246,CUST007478,1994-06-17,2027-03-10,Canada,Cancelled
444,CUST006728,1987-11-11,2027-03-10,USA,Active
743,CUST005767,1966-01-30,2027-03-10,USA,Active
833,CUST009022,1996-06-16,2027-03-10,Singapore,Active
1639,CUST000380,1964-10-18,2027-03-10,USA,Cancelled
2733,CUST009879,1958-07-11,2027-03-10,Germany,Active
4145,CUST004300,1964-10-29,2027-03-10,USA,Active
4624,CUST010028,1972-03-12,2027-03-10,Germany,Active
4700,CUST000388,1969-01-29,2027-03-10,India,Inactive
5450,CUST008716,1972-03-12,2027-03-10,Germany,Active


In [50]:
#Future Signup Date: If the date is in the future and no authoritative source can establish the correct date,
#set it to missing and flag it as invalid.
#Do not substitute today's date or guess a date.
future_signup = df["Signup_Date"] > pd.Timestamp.today()

df["Signup_Quality"] = "Valid"

df.loc[df["Signup_Date"].isna(), "Signup_Quality"] = "Missing"
df.loc[future_signup, "Signup_Quality"] = "Invalid_Future"

df["Signup_Quality"].value_counts()

,count
Signup_Quality,
Valid,10079
Invalid_Future,21


In [51]:
df.loc[future_signup, "Signup_Date"] = pd.NaT

In [52]:
invalid_date_order = (
    df["Date_of_Birth"].notna()
    & df["Signup_Date"].notna()
    & (df["Date_of_Birth"] >= df["Signup_Date"])
)

invalid_date_order.sum()

np.int64(0)

In [53]:
df["DOB_Quality"].value_counts(dropna=False)

,count
DOB_Quality,
Valid,9921
Missing,154
Invalid_Future,25


In [54]:
df["Signup_Quality"].value_counts(dropna=False)

,count
Signup_Quality,
Valid,10079
Invalid_Future,21


In [55]:
# MISSING COLUMNS

missing_summary = pd.DataFrame({
    "Missing_Count": df.isna().sum(),
    "Missing_Percentage": (df.isna().mean() * 100).round(2)
})

missing_summary.sort_values("Missing_Count", ascending=False)

,Missing_Count,Missing_Percentage
Phone,255,2.52
Phone_Digits,255,2.52
Date_of_Birth,179,1.77
Email,118,1.17
Email_Normalized,118,1.17
State,102,1.01
First_Name,80,0.79
Signup_Date,21,0.21
Country,0,0.00
Last_Name,0,0.00


In [56]:
# Tackle Missing Customer First Name

df.loc[
    df["First_Name"].isna(),
    ["Customer_ID", "Last_Name", "Email", "Phone", "Country", "State", "Signup_Date", "Customer_Status"]
].head(20)

,Customer_ID,Last_Name,Email,Phone,Country,State,Signup_Date,Customer_Status
192,CUST002274,Mehta,noah.mehta3274@example.com,+61-473231538,Australia,New South Wales,2022-03-02,Active
325,CUST009012,Singh,sarah.singh10012@example.com,+1-5625417153,USA,Texas,2024-02-18,Active
683,CUST007926,Shah,william.shah8926@example.com,919746426466,India,Karnataka,2022-11-08,Active
715,CUST004470,Johnson,sophia.johnson5470@example.com,14345964810,USA,Florida,2019-02-26,Inactive
764,CUST007723,Clark,zara.clark8723@example.com,+65-94093974,Singapore,North Region,2023-08-26,Active
958,CUST000211,Wilson,karan.wilson1211@example.com,+49-1551108979,Germany,Berlin,2023-05-27,Inactive
965,CUST009286,Thomas,william.thomas10286@example.com,+1-5606415195,USA,New York,2022-04-02,Active
1175,CUST008782,Miller,karan.miller9782@example.com,+1-7019235919,USA,Florida,2023-08-14,Active
1310,CUST008230,Shah,meera.shah9230@example.com,+1-6506041519,USA,Florida,2020-03-09,Active
1525,CUST005299,Clark,alex.clark6299@example.com,+49-1767358908,Germany,Bavaria,2020-08-06,Active


In [57]:
df["First_Name_Quality"] = "Valid"

df.loc[
    df["First_Name"].isna(),
    "First_Name_Quality"
] = "Missing"

df["First_Name_Quality"].value_counts()

,count
First_Name_Quality,
Valid,10020
Missing,80


In [58]:
# Tackle Missing State

df["State"].value_counts(dropna=False)

,count
State,
Delhi,489
Karnataka,486
Gujarat,480
Maharashtra,479
Telangana,477
Washington,450
Tamil Nadu,449
Texas,448
Illinois,433


In [59]:
pd.crosstab(
    df["Country"],
    df["State"],
    dropna=False
)

State,Alberta,Bavaria,Berlin,British Columbia,California,Central Region,Delhi,East Region,England,Florida,...,Scotland,Tamil Nadu,Telangana,Texas,Victoria,Wales,Washington,West Region,Western Australia,NaN
Country,,,,,,,,,,,,,,,,,,,,,
Australia,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,196,0,0,0,214,8
Canada,231,0,0,302,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,11
Germany,0,195,197,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,12
India,0,0,0,0,0,0,489,0,0,0,...,0,449,477,0,0,0,0,0,0,29
Singapore,0,0,0,0,0,189,0,182,0,0,...,0,0,0,0,0,0,0,172,0,8
UK,0,0,0,0,0,0,0,0,294,0,...,303,0,0,0,0,301,0,0,0,12
USA,0,0,0,0,429,0,0,0,0,389,...,0,0,0,448,0,0,450,0,0,22


In [60]:
df["State_Quality"] = "Valid"

df.loc[
    df["State"].isna(),
    "State_Quality"
] = "Missing"

df["State_Quality"].value_counts()

,count
State_Quality,
Valid,9998
Missing,102


In [61]:
#FINAL VALIDATION

print("Rows:", len(df))
print("Columns:", len(df.columns))
print("Exact duplicate rows:", df.duplicated().sum())

print("Missing Customer_ID:", df["Customer_ID"].isna().sum())
print("Duplicate Customer_ID:", df["Customer_ID"].duplicated().sum())

Rows: 10100
Columns: 18
Exact duplicate rows: 0
Missing Customer_ID: 0
Duplicate Customer_ID: 0


In [62]:
print(df["Country"].value_counts(dropna=False))
print()
print(df["Customer_Status"].value_counts(dropna=False))

Country
India        2889
USA          2553
UK           1230
Canada       1052
Germany       819
Australia     807
Singapore     750
Name: count, dtype: int64

Customer_Status
Active       6162
Inactive     2078
Cancelled    1261
Suspended     599
Name: count, dtype: int64


In [63]:
phone_check = pd.crosstab(
    df["Country"],
    df["Phone_Digits"].str.len(),
    dropna=False
)

phone_check

Phone_Digits,10,11,12,<NA>
Country,,,,
Australia,0,782,0,25
Canada,0,1025,0,27
Germany,0,0,797,22
India,0,0,2819,70
Singapore,733,0,0,17
UK,0,0,1196,34
USA,0,2493,0,60


In [64]:
df["Email_Quality"].value_counts()

,count
Email_Quality,
Valid,9860
Invalid,122
Missing,118


In [65]:
missing_summary = pd.DataFrame({
    "Missing_Count": df.isna().sum(),
    "Missing_Percentage": (df.isna().mean() * 100).round(2)
})

missing_summary.sort_values("Missing_Count", ascending=False)

,Missing_Count,Missing_Percentage
Phone_Digits,255,2.52
Phone,255,2.52
Date_of_Birth,179,1.77
Email,118,1.17
Email_Normalized,118,1.17
State,102,1.01
First_Name,80,0.79
Signup_Date,21,0.21
Customer_ID,0,0.00
Last_Name,0,0.00


In [66]:
print("Rows:", len(df))
print("Columns:", len(df.columns))
print("Exact duplicate rows:", df.duplicated().sum())

print("Missing Customer_ID:", df["Customer_ID"].isna().sum())
print("Duplicate Customer_ID:", df["Customer_ID"].duplicated().sum())

Rows: 10100
Columns: 18
Exact duplicate rows: 0
Missing Customer_ID: 0
Duplicate Customer_ID: 0


In [67]:
print("COUNTRIES")
print(df["Country"].value_counts(dropna=False))

print("\nCUSTOMER STATUS")
print(df["Customer_Status"].value_counts(dropna=False))

COUNTRIES
Country
India        2889
USA          2553
UK           1230
Canada       1052
Germany       819
Australia     807
Singapore     750
Name: count, dtype: int64

CUSTOMER STATUS
Customer_Status
Active       6162
Inactive     2078
Cancelled    1261
Suspended     599
Name: count, dtype: int64


In [68]:
# FINAL QUALITY CHECKS

#Step 1: Validate the date-quality flags
print("DOB QUALITY")
print(df["DOB_Quality"].value_counts(dropna=False))

print("\nSIGNUP DATE QUALITY")
print(df["Signup_Quality"].value_counts(dropna=False))

DOB QUALITY
DOB_Quality
Valid             9921
Missing            154
Invalid_Future      25
Name: count, dtype: int64

SIGNUP DATE QUALITY
Signup_Quality
Valid             10079
Invalid_Future       21
Name: count, dtype: int64


In [69]:
print("Future DOBs remaining:",
      (df["Date_of_Birth"] > pd.Timestamp.today()).sum())

print("Future Signup Dates remaining:",
      (df["Signup_Date"] > pd.Timestamp.today()).sum())

Future DOBs remaining: 0
Future Signup Dates remaining: 0


In [70]:
#Step 2: Check that our important source fields weren't accidentally altered

df[[
    "Customer_ID",
    "First_Name",
    "Last_Name",
    "Email",
    "Phone",
    "Date_of_Birth",
    "Country",
    "State",
    "Signup_Date",
    "Customer_Status"
]].info()

<class 'pandas.core.frame.DataFrame'>
Index: 10100 entries, 0 to 10179
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Customer_ID      10100 non-null  object        
 1   First_Name       10020 non-null  object        
 2   Last_Name        10100 non-null  object        
 3   Email            9982 non-null   object        
 4   Phone            9845 non-null   object        
 5   Date_of_Birth    9921 non-null   datetime64[ns]
 6   Country          10100 non-null  object        
 7   State            9998 non-null   object        
 8   Signup_Date      10079 non-null  datetime64[ns]
 9   Customer_Status  10100 non-null  object        
dtypes: datetime64[ns](2), object(8)
memory usage: 868.0+ KB


In [71]:
#final data-quality summary

quality_summary = pd.DataFrame({
    "Missing_Count": df.isna().sum(),
    "Missing_Percentage": (df.isna().mean() * 100).round(2)
})

quality_summary = quality_summary[
    quality_summary["Missing_Count"] > 0
].sort_values("Missing_Count", ascending=False)

quality_summary

,Missing_Count,Missing_Percentage
Phone,255,2.52
Phone_Digits,255,2.52
Date_of_Birth,179,1.77
Email,118,1.17
Email_Normalized,118,1.17
State,102,1.01
First_Name,80,0.79
Signup_Date,21,0.21


In [72]:
# SAVE CSV

df.to_csv(folder_path+'customer_data_cleaned.csv')
df.shape

(10100, 18)

In [73]:
# DOCUMENTATION

quality_report = pd.DataFrame([
    ["Exact duplicate rows", 80, "Removed", "Redundant records"],
    ["Duplicate Customer_ID", 0, "No action", "Customer_ID is unique"],
    ["Missing Phone", 255, "Retained", "No reliable value available"],
    ["Missing Email", 118, "Retained", "No reliable value available"],
    ["Invalid Email", 122, "Flagged", "Do not fabricate/correct without source"],
    ["Future DOB", 25, "Set to missing", "Invalid future date"],
    ["Missing DOB", 154, "Retained", "Missing allowed; no reliable replacement"],
    ["Missing State", 102, "Retained", "Cannot reliably infer state"],
    ["Missing First Name", 80, "Retained", "Cannot reliably infer from email"],
    ["Future Signup Date", 21, "Set to missing", "Invalid future date"],
    ["DOB >= Signup Date", 0, "No action", "No chronological violations"]
], columns=["Issue", "Records", "Treatment", "Business_Rationale"])

quality_report

,Issue,Records,Treatment,Business_Rationale
0,Exact duplicate rows,80,Removed,Redundant records
1,Duplicate Customer_ID,0,No action,Customer_ID is unique
2,Missing Phone,255,Retained,No reliable value available
3,Missing Email,118,Retained,No reliable value available
4,Invalid Email,122,Flagged,Do not fabricate/correct without source
5,Future DOB,25,Set to missing,Invalid future date
6,Missing DOB,154,Retained,Missing allowed; no reliable replacement
7,Missing State,102,Retained,Cannot reliably infer state
8,Missing First Name,80,Retained,Cannot reliably infer from email
9,Future Signup Date,21,Set to missing,Invalid future date


# EDA

In [74]:
# How is the customer base distributed across countries?
country_summary = (
    df["Country"]
    .value_counts()
    .rename_axis("Country")
    .reset_index(name="Customer_Count")
)

country_summary

,Country,Customer_Count
0,India,2889
1,USA,2553
2,UK,1230
3,Canada,1052
4,Germany,819
5,Australia,807
6,Singapore,750


In [75]:
country_summary["Percentage"] = (
    country_summary["Customer_Count"]
    / len(df) * 100
).round(2)

country_summary

,Country,Customer_Count,Percentage
0,India,2889,28.60
1,USA,2553,25.28
2,UK,1230,12.18
3,Canada,1052,10.42
4,Germany,819,8.11
5,Australia,807,7.99
6,Singapore,750,7.43



> India and USA together account for 53.88% of the customer base, making them the two largest customer markets. The UK is the third-largest market at 12.18%.




In [76]:
status_summary = (
    df["Customer_Status"]
    .value_counts()
    .rename_axis("Customer_Status")
    .reset_index(name="Customer_Count")
)

status_summary["Percentage"] = (
    status_summary["Customer_Count"] / len(df) * 100
).round(2)

status_summary

,Customer_Status,Customer_Count,Percentage
0,Active,6162,61.01
1,Inactive,2078,20.57
2,Cancelled,1261,12.49
3,Suspended,599,5.93


61.01% of customers are Active, while 38.99% are not currently Active (Inactive, Cancelled, or Suspended).

In [77]:
#Does customer status differ by country?
status_by_country = pd.crosstab(
    df["Country"],
    df["Customer_Status"],
    normalize="index"
).mul(100).round(2)

status_by_country

Customer_Status,Active,Cancelled,Inactive,Suspended
Country,,,,
Australia,59.73,13.26,19.21,7.81
Canada,59.22,14.26,20.44,6.08
Germany,62.39,9.77,22.34,5.49
India,61.13,12.56,20.77,5.54
Singapore,62.67,10.40,21.87,5.07
UK,61.38,13.66,19.43,5.53
USA,60.91,12.34,20.45,6.31


The interesting observations

1. Active customers are remarkably similar across countries.

They range from about 59.2% to 62.7%.

So I would not make a dramatic claim such as "Singapore has significantly more active customers." The difference is only ~3.5 percentage points.

2. Canada has the highest cancellation percentage: 14.26%.

That's worth flagging for further investigation.

3. Australia has the highest suspension percentage: 7.81%.

Again, interesting, but we don't know why. Don't invent a business explanation.

4. Germany has the highest inactive percentage: 22.34%.

That could be worth investigating, particularly if "Inactive" has a meaningful business definition.

5. Singapore has the highest active percentage: 62.67%.

But the difference isn't large enough by itself to conclude that Singapore is a better-performing market.

"Canada has the highest proportion of customers classified as Cancelled (14.26%), while Germany has the lowest (9.77%). This difference warrants further investigation, but customer status alone is insufficient to infer retention or churn performance."

In [78]:
#Are newer customers more likely to be Active, Inactive, Cancelled, or Suspended?

df["Signup_Year"] = df["Signup_Date"].dt.year

pd.crosstab(
    df["Signup_Year"],
    df["Customer_Status"],
    normalize="index"
).mul(100).round(2)

Customer_Status,Active,Cancelled,Inactive,Suspended
Signup_Year,,,,
2019.0,60.04,12.70,20.78,6.48
2020.0,62.17,12.17,19.77,5.89
2021.0,62.04,13.09,19.70,5.17
2022.0,59.96,11.40,22.56,6.08
2023.0,61.41,13.62,18.55,6.42
2024.0,60.17,12.49,21.56,5.78
2025.0,61.12,11.44,21.86,5.59


Customers who signed up in 2022 have the lowest Active proportion and the highest Inactive proportion among the signup cohorts shown. This cohort may warrant further investigation.

In [79]:
# Next: let's investigate the actual data-quality issues
#Do different Customer_IDs appear to represent the same customer?
# Step 1: duplicate normalized emails

email_dup_mask = (
    df["Email_Normalized"].notna()
    & df["Email_Normalized"].duplicated(keep=False)
)

email_dup_mask.sum()

np.int64(318)

In [80]:
df[email_dup_mask].sort_values("Email_Normalized")

,Customer_ID,First_Name,Last_Name,Email,Phone,Date_of_Birth,Country,State,Signup_Date,Customer_Status,Email_Normalized,Phone_Digits,Email_Valid,Email_Quality,DOB_Quality,Signup_Quality,First_Name_Quality,State_Quality,Signup_Year
6262,CUST003976,Aarav,Gupta,aarav.gupta4976@example.com,+91-8534142254,2000-07-24,India,Gujarat,2024-06-16,Active,aarav.gupta4976@example.com,918534142254,True,Valid,Valid,Valid,Valid,Valid,2024.0
5568,CUST010063,Aarav,Gupta,aarav.gupta4976@example.com,+91-8534142254,2000-07-24,India,Gujarat,2024-06-16,Active,aarav.gupta4976@example.com,918534142254,True,Valid,Valid,Valid,Valid,Valid,2024.0
112,CUST010044,Aarav,Harris,aarav.harris10768@example.com,+1-7865848061,1976-03-12,USA,Washington,2024-04-14,Active,aarav.harris10768@example.com,17865848061,True,Valid,Valid,Valid,Valid,Valid,2024.0
5174,CUST009768,Aarav,Harris,aarav.harris10768@example.com,+1-7865848061,1976-03-12,USA,Washington,2024-04-14,Active,aarav.harris10768@example.com,17865848061,True,Valid,Valid,Valid,Valid,Valid,2024.0
5917,CUST004740,Aarav,Johnson,aarav.johnson5740@example.com,+65-87866323,1991-07-12,Singapore,Central Region,2020-12-11,Active,aarav.johnson5740@example.com,6587866323,True,Valid,Valid,Valid,Valid,Valid,2020.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6143,CUST009809,William,Miller,william.miller10809@example.com,+91-6417633905,1977-08-07,India,Karnataka,2025-05-01,Inactive,william.miller10809@example.com,916417633905,True,Valid,Valid,Valid,Valid,Valid,2025.0
9801,CUST005316,zara,Anderson,zara.anderson6316@example.com,+1-8959487461,1962-11-26,USA,Washington,2019-12-28,Active,zara.anderson6316@example.com,18959487461,True,Valid,Valid,Valid,Valid,Valid,2019.0
5172,CUST010043,zara,Anderson,zara.anderson6316@example.com,+1-8959487461,1962-11-26,USA,Washington,2019-12-28,Active,zara.anderson6316@example.com,18959487461,True,Valid,Valid,Valid,Valid,Valid,2019.0
8301,CUST010033,Zara,THOMAS,zara.thomas4932@example.com,+1-9889867251,1995-08-31,USA,Illinois,2019-01-21,Inactive,zara.thomas4932@example.com,19889867251,True,Valid,Valid,Valid,Valid,Valid,2019.0


In [81]:
df[email_dup_mask]["Email_Normalized"].nunique()

103

In [82]:
df[email_dup_mask].groupby("Email_Normalized").size().value_counts().sort_index()

,count
2,98
18,1
22,1
25,1
28,1
29,1


In [83]:
df[email_dup_mask].groupby("Email_Normalized").size().sort_values(ascending=False).head(10)

,0
Email_Normalized,
user@,29
not_an_email,28
user.example.com,25
missing_at_symbol.com,22
john..smith@example.com,18
aarav.harris10768@example.com,2
aarav.gupta4976@example.com,2
alex.anderson8016@example.com,2
alex.chen5734@example.com,2


In [84]:
valid_email_dup = (
    email_dup_mask
    & df["Email_Quality"].eq("Valid")
)

valid_email_dup.sum()

np.int64(196)

In [85]:
df[valid_email_dup].sort_values("Email_Normalized")

,Customer_ID,First_Name,Last_Name,Email,Phone,Date_of_Birth,Country,State,Signup_Date,Customer_Status,Email_Normalized,Phone_Digits,Email_Valid,Email_Quality,DOB_Quality,Signup_Quality,First_Name_Quality,State_Quality,Signup_Year
5568,CUST010063,Aarav,Gupta,aarav.gupta4976@example.com,+91-8534142254,2000-07-24,India,Gujarat,2024-06-16,Active,aarav.gupta4976@example.com,918534142254,True,Valid,Valid,Valid,Valid,Valid,2024.0
6262,CUST003976,Aarav,Gupta,aarav.gupta4976@example.com,+91-8534142254,2000-07-24,India,Gujarat,2024-06-16,Active,aarav.gupta4976@example.com,918534142254,True,Valid,Valid,Valid,Valid,Valid,2024.0
112,CUST010044,Aarav,Harris,aarav.harris10768@example.com,+1-7865848061,1976-03-12,USA,Washington,2024-04-14,Active,aarav.harris10768@example.com,17865848061,True,Valid,Valid,Valid,Valid,Valid,2024.0
5174,CUST009768,Aarav,Harris,aarav.harris10768@example.com,+1-7865848061,1976-03-12,USA,Washington,2024-04-14,Active,aarav.harris10768@example.com,17865848061,True,Valid,Valid,Valid,Valid,Valid,2024.0
5917,CUST004740,Aarav,Johnson,aarav.johnson5740@example.com,+65-87866323,1991-07-12,Singapore,Central Region,2020-12-11,Active,aarav.johnson5740@example.com,6587866323,True,Valid,Valid,Valid,Valid,Valid,2020.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6143,CUST009809,William,Miller,william.miller10809@example.com,+91-6417633905,1977-08-07,India,Karnataka,2025-05-01,Inactive,william.miller10809@example.com,916417633905,True,Valid,Valid,Valid,Valid,Valid,2025.0
5172,CUST010043,zara,Anderson,zara.anderson6316@example.com,+1-8959487461,1962-11-26,USA,Washington,2019-12-28,Active,zara.anderson6316@example.com,18959487461,True,Valid,Valid,Valid,Valid,Valid,2019.0
9801,CUST005316,zara,Anderson,zara.anderson6316@example.com,+1-8959487461,1962-11-26,USA,Washington,2019-12-28,Active,zara.anderson6316@example.com,18959487461,True,Valid,Valid,Valid,Valid,Valid,2019.0
2836,CUST003932,Zara,THOMAS,zara.thomas4932@example.com,+1-9889867251,1995-08-31,USA,Illinois,2019-01-21,Inactive,zara.thomas4932@example.com,19889867251,True,Valid,Valid,Valid,Valid,Valid,2019.0


In [86]:
df.loc[
    df["Email_Normalized"].str.contains(r"\.\.", regex=True, na=False),
    "Email_Quality"
].value_counts()

,count
Email_Quality,
Invalid,18


In [87]:
df.loc[
    df["Email_Normalized"].str.contains(r"\.\.", regex=True, na=False),
    ["Customer_ID", "Email", "Email_Normalized", "Email_Quality"]
]

,Customer_ID,Email,Email_Normalized,Email_Quality
291,CUST007738,john..smith@example.com,john..smith@example.com,Invalid
2138,CUST005014,john..smith@example.com,john..smith@example.com,Invalid
2473,CUST008331,john..smith@example.com,john..smith@example.com,Invalid
2645,CUST007387,john..smith@example.com,john..smith@example.com,Invalid
3125,CUST001283,john..smith@example.com,john..smith@example.com,Invalid
3200,CUST001840,john..smith@example.com,john..smith@example.com,Invalid
3424,CUST003338,john..smith@example.com,john..smith@example.com,Invalid
3647,CUST002500,john..smith@example.com,john..smith@example.com,Invalid
4010,CUST008928,john..smith@example.com,john..smith@example.com,Invalid
4165,CUST006706,john..smith@example.com,john..smith@example.com,Invalid


In [88]:
df.loc[
    df["Email_Quality"] == "Invalid",
    ["Customer_ID", "Email", "Email_Normalized"]
].head(20)

,Customer_ID,Email,Email_Normalized
12,CUST005921,user.example.com,user.example.com
35,CUST004922,user.example.com,user.example.com
291,CUST007738,john..smith@example.com,john..smith@example.com
513,CUST000089,missing_at_symbol.com,missing_at_symbol.com
812,CUST007385,not_an_email,not_an_email
967,CUST001459,user.example.com,user.example.com
975,CUST005466,user@,user@
985,CUST010065,not_an_email,not_an_email
1279,CUST000701,not_an_email,not_an_email
1328,CUST003686,user@,user@


In [89]:
df.loc[
    df["Email_Quality"] == "Invalid",
    "Email_Normalized"
].value_counts().head(20)

,count
Email_Normalized,
user@,29
not_an_email,28
user.example.com,25
missing_at_symbol.com,22
john..smith@example.com,18


Invalid Email: 122 records flagged. Invalid patterns include incomplete addresses, missing @ symbols, malformed domains, and consecutive dots. Records were retained because no authoritative source was available to safely correct the values.


In [90]:
#DUPLICATE EMAILS
#Are these duplicate emails actually duplicate customers, or are they separate customer records sharing an email?
df[df["Email_Normalized"].duplicated(keep=False)].groupby(
    "Email_Normalized"
)["Customer_ID"].nunique().value_counts()

,count
Customer_ID,
2,98
18,1
22,1
28,1
25,1
29,1


The 5 large groups are the interesting part

We already saw examples like:

user@ → 29 records
not_an_email → 28
user.example.com → 25
missing_at_symbol.com → 22
john..smith@example.com → 18

These are invalid email patterns, not genuine shared customer emails.

So we should not treat those 122 records as customer duplication. We've already classified them as invalid email data.

That leaves the 98 pairs as the real duplicate-email cases we need to investigate.

In [91]:
valid_email_dup_mask = (
    (df["Email_Quality"] == "Valid") &
    (df["Email_Normalized"].notna()) &
    (df.groupby("Email_Normalized")["Email_Normalized"].transform("size") == 2)
)

valid_email_dups = df.loc[
    valid_email_dup_mask,
    ["Customer_ID", "First_Name", "Last_Name",
     "Email", "Email_Normalized", "Phone_Digits",
     "Date_of_Birth", "Country", "State",
     "Signup_Date", "Customer_Status"]
].sort_values("Email_Normalized")

valid_email_dups

,Customer_ID,First_Name,Last_Name,Email,Email_Normalized,Phone_Digits,Date_of_Birth,Country,State,Signup_Date,Customer_Status
5568,CUST010063,Aarav,Gupta,aarav.gupta4976@example.com,aarav.gupta4976@example.com,918534142254,2000-07-24,India,Gujarat,2024-06-16,Active
6262,CUST003976,Aarav,Gupta,aarav.gupta4976@example.com,aarav.gupta4976@example.com,918534142254,2000-07-24,India,Gujarat,2024-06-16,Active
112,CUST010044,Aarav,Harris,aarav.harris10768@example.com,aarav.harris10768@example.com,17865848061,1976-03-12,USA,Washington,2024-04-14,Active
5174,CUST009768,Aarav,Harris,aarav.harris10768@example.com,aarav.harris10768@example.com,17865848061,1976-03-12,USA,Washington,2024-04-14,Active
5917,CUST004740,Aarav,Johnson,aarav.johnson5740@example.com,aarav.johnson5740@example.com,6587866323,1991-07-12,Singapore,Central Region,2020-12-11,Active
...,...,...,...,...,...,...,...,...,...,...,...
6143,CUST009809,William,Miller,william.miller10809@example.com,william.miller10809@example.com,916417633905,1977-08-07,India,Karnataka,2025-05-01,Inactive
5172,CUST010043,zara,Anderson,zara.anderson6316@example.com,zara.anderson6316@example.com,18959487461,1962-11-26,USA,Washington,2019-12-28,Active
9801,CUST005316,zara,Anderson,zara.anderson6316@example.com,zara.anderson6316@example.com,18959487461,1962-11-26,USA,Washington,2019-12-28,Active
2836,CUST003932,Zara,THOMAS,zara.thomas4932@example.com,zara.thomas4932@example.com,19889867251,1995-08-31,USA,Illinois,2019-01-21,Inactive


In [92]:
#Compare the 98 pairs of duplicated email records to see if they are the same customer.

match_cols = [
    "First_Name",
    "Last_Name",
    "Phone_Digits",
    "Date_of_Birth",
    "Country",
    "State",
    "Signup_Date",
    "Customer_Status"
]

duplicate_records = df.loc[
    df["Email_Normalized"].duplicated(keep=False)
    & df["Email_Normalized"].notna()
    & (df["Email_Quality"] == "Valid")
].copy()

# Compare each field within an email group
for col in match_cols:
    duplicate_records[f"{col}_Match"] = (
        duplicate_records.groupby("Email_Normalized")[col]
        .transform(lambda x: x.nunique(dropna=False) == 1)
    )

duplicate_records["All_Attributes_Match"] = duplicate_records[
    [f"{col}_Match" for col in match_cols]
].all(axis=1)

duplicate_records[
    ["Customer_ID", "Email_Normalized"] +
    [f"{col}_Match" for col in match_cols] +
    ["All_Attributes_Match"]
].head(20)

,Customer_ID,Email_Normalized,First_Name_Match,Last_Name_Match,Phone_Digits_Match,Date_of_Birth_Match,Country_Match,State_Match,Signup_Date_Match,Customer_Status_Match,All_Attributes_Match
76,CUST006573,arjun.smith7573@example.com,True,True,True,True,True,True,True,True,True
112,CUST010044,aarav.harris10768@example.com,True,True,True,True,True,True,True,True,True
231,CUST008760,isabella.kumar9760@example.com,True,True,True,True,True,True,True,True,True
286,CUST006056,thomas.martin7056@example.com,True,True,True,True,True,True,True,True,True
289,CUST004606,james.shah5606@example.com,True,True,True,True,True,True,True,True,True
316,CUST008193,thomas.khan9193@example.com,True,True,True,True,True,True,True,True,True
318,CUST001072,john.miller2072@example.com,True,True,True,True,True,True,True,True,True
372,CUST001671,chloe.roberts2671@example.com,True,True,True,True,True,True,True,True,True
426,CUST009431,priya.taylor10431@example.com,True,True,True,True,True,True,True,True,True
474,CUST010064,michael.brown5479@example.com,True,True,True,True,True,True,True,True,True


In [93]:
pair_analysis = (
    duplicate_records
    .groupby("Email_Normalized")
    .agg(
        Records=("Customer_ID", "size"),
        Customer_IDs=("Customer_ID", lambda x: list(x)),
        First_Name=("First_Name", lambda x: x.nunique(dropna=False)),
        Last_Name=("Last_Name", lambda x: x.nunique(dropna=False)),
        Phone=("Phone_Digits", lambda x: x.nunique(dropna=False)),
        DOB=("Date_of_Birth", lambda x: x.nunique(dropna=False)),
        Country=("Country", lambda x: x.nunique(dropna=False)),
        State=("State", lambda x: x.nunique(dropna=False)),
        Signup_Date=("Signup_Date", lambda x: x.nunique(dropna=False)),
        Status=("Customer_Status", lambda x: x.nunique(dropna=False))
    )
    .reset_index()
)

pair_analysis.head(20)

,Email_Normalized,Records,Customer_IDs,First_Name,Last_Name,Phone,DOB,Country,State,Signup_Date,Status
0,aarav.gupta4976@example.com,2,"[CUST010063, CUST003976]",1,1,1,1,1,1,1,1
1,aarav.harris10768@example.com,2,"[CUST010044, CUST009768]",1,1,1,1,1,1,1,1
2,aarav.johnson5740@example.com,2,"[CUST004740, CUST010083]",1,1,1,1,1,1,1,1
3,aarav.kumar4269@example.com,2,"[CUST010042, CUST003269]",1,1,1,1,1,1,1,1
4,aarav.thomas4010@example.com,2,"[CUST010035, CUST003010]",1,1,1,1,1,1,1,1
5,aisha.anderson5550@example.com,2,"[CUST004550, CUST010087]",1,1,1,1,1,1,1,1
6,aisha.williams3017@example.com,2,"[CUST010014, CUST002017]",1,1,1,1,1,1,1,1
7,alex.anderson8016@example.com,2,"[CUST007016, CUST010001]",1,1,1,1,1,1,1,1
8,alex.chen5734@example.com,2,"[CUST004734, CUST010054]",1,1,1,1,1,1,1,1
9,alex.kumar3695@example.com,2,"[CUST010081, CUST002695]",1,1,1,1,1,1,1,1


In [94]:
pair_analysis["Exact_Match"] = (
    (pair_analysis["Records"] == 2) &
    (pair_analysis["First_Name"] == 1) &
    (pair_analysis["Last_Name"] == 1) &
    (pair_analysis["Phone"] == 1) &
    (pair_analysis["DOB"] == 1) &
    (pair_analysis["Country"] == 1) &
    (pair_analysis["State"] == 1) &
    (pair_analysis["Signup_Date"] == 1) &
    (pair_analysis["Status"] == 1)
)

pair_analysis["Exact_Match"].value_counts()


,count
Exact_Match,
True,98


So we have strong evidence that these are true duplicate customer records, not legitimate cases of two customers sharing an email.

Which Customer_ID becomes the surviving/master record, and what happens to the other Customer_ID?

In [95]:
pair_analysis["Customer_IDs"].head(20)

,Customer_IDs
0,"[CUST010063, CUST003976]"
1,"[CUST010044, CUST009768]"
2,"[CUST004740, CUST010083]"
3,"[CUST010042, CUST003269]"
4,"[CUST010035, CUST003010]"
5,"[CUST004550, CUST010087]"
6,"[CUST010014, CUST002017]"
7,"[CUST007016, CUST010001]"
8,"[CUST004734, CUST010054]"
9,"[CUST010081, CUST002695]"


In [96]:
duplicate_records[
    duplicate_records["Customer_ID"].str.startswith("CUST010")
]["Customer_ID"].count()

np.int64(98)

In [97]:
# Compare all non-ID attributes within each duplicate pair

duplicate_records_clean = duplicate_records.reset_index(drop=True)

comparison_cols = [
    "First_Name",
    "Last_Name",
    "Phone_Digits",
    "Date_of_Birth",
    "Country",
    "State",
    "Signup_Date",
    "Customer_Status"
]

attribute_comparison = (
    duplicate_records_clean
    .groupby("Email_Normalized")[comparison_cols]
    .nunique(dropna=False)
)

attribute_comparison.head(10)


,First_Name,Last_Name,Phone_Digits,Date_of_Birth,Country,State,Signup_Date,Customer_Status
Email_Normalized,,,,,,,,
aarav.gupta4976@example.com,1,1,1,1,1,1,1,1
aarav.harris10768@example.com,1,1,1,1,1,1,1,1
aarav.johnson5740@example.com,1,1,1,1,1,1,1,1
aarav.kumar4269@example.com,1,1,1,1,1,1,1,1
aarav.thomas4010@example.com,1,1,1,1,1,1,1,1
aisha.anderson5550@example.com,1,1,1,1,1,1,1,1
aisha.williams3017@example.com,1,1,1,1,1,1,1,1
alex.anderson8016@example.com,1,1,1,1,1,1,1,1
alex.chen5734@example.com,1,1,1,1,1,1,1,1


In [98]:
attribute_comparison.value_counts()

,,,,,,,,count
First_Name,Last_Name,Phone_Digits,Date_of_Birth,Country,State,Signup_Date,Customer_Status,
1,1,1,1,1,1,1,1,98


What we've established

We started with:

98 duplicate email pairs
196 records
Each pair has one CUST010xxx record and one pre-existing customer record
Email is valid and duplicated exactly twice
After normalization, the emails match
Now we've compared the other customer attributes

And the result is:

98/98 pairs have complete attribute agreement.

So these aren't merely "possible duplicates". They are confirmed exact duplicate records.

For example:

CUST010063 ↔ CUST003976

has identical:

Name
Phone
DOB
Country
State
Signup date
Status
Email

Same for all 98 pairs.

In [99]:
#Next step: create the duplicate mapping
#Which Customer_ID becomes the surviving/master ID?
#Retain the non-CUST010xxx Customer_ID as the master and mark the CUST010xxx record as the duplicate.

pair_analysis[["Customer_IDs"]].head(10)

,Customer_IDs
0,"[CUST010063, CUST003976]"
1,"[CUST010044, CUST009768]"
2,"[CUST004740, CUST010083]"
3,"[CUST010042, CUST003269]"
4,"[CUST010035, CUST003010]"
5,"[CUST004550, CUST010087]"
6,"[CUST010014, CUST002017]"
7,"[CUST007016, CUST010001]"
8,"[CUST004734, CUST010054]"
9,"[CUST010081, CUST002695]"


In [100]:
duplicate_mapping = pd.DataFrame({
    "Duplicate_Customer_ID": pair_analysis["Customer_IDs"].apply(
        lambda x: next(cid for cid in x if cid.startswith("CUST010"))
    ),
    "Master_Customer_ID": pair_analysis["Customer_IDs"].apply(
        lambda x: next(cid for cid in x if not cid.startswith("CUST010"))
    )
})

duplicate_mapping.head(10)

,Duplicate_Customer_ID,Master_Customer_ID
0,CUST010063,CUST003976
1,CUST010044,CUST009768
2,CUST010083,CUST004740
3,CUST010042,CUST003269
4,CUST010035,CUST003010
5,CUST010087,CUST004550
6,CUST010014,CUST002017
7,CUST010001,CUST007016
8,CUST010054,CUST004734
9,CUST010081,CUST002695


In [101]:
duplicate_mapping.shape

(98, 2)

In [102]:
duplicate_mapping["Duplicate_Customer_ID"].str.startswith("CUST010").value_counts()

,count
Duplicate_Customer_ID,
True,98


In [103]:
duplicate_mapping["Master_Customer_ID"].str.startswith("CUST010").value_counts()

,count
Master_Customer_ID,
False,98


In [104]:
#Creating Golden Record

golden_record_map = pair_analysis[["Customer_IDs"]].copy()

golden_record_map["Master_Customer_ID"] = [
    f"MASTER{i:04d}" for i in range(1, len(golden_record_map) + 1)
]

golden_record_map = golden_record_map[
    ["Master_Customer_ID", "Customer_IDs"]
]

golden_record_map.head(10)



,Master_Customer_ID,Customer_IDs
0,MASTER0001,"[CUST010063, CUST003976]"
1,MASTER0002,"[CUST010044, CUST009768]"
2,MASTER0003,"[CUST004740, CUST010083]"
3,MASTER0004,"[CUST010042, CUST003269]"
4,MASTER0005,"[CUST010035, CUST003010]"
5,MASTER0006,"[CUST004550, CUST010087]"
6,MASTER0007,"[CUST010014, CUST002017]"
7,MASTER0008,"[CUST007016, CUST010001]"
8,MASTER0009,"[CUST004734, CUST010054]"
9,MASTER0010,"[CUST010081, CUST002695]"


In [105]:
golden_record_map.shape

(98, 2)

In [106]:
# Get all Customer IDs involved in duplicate pairs
duplicate_customer_ids = [
    customer_id
    for ids in pair_analysis["Customer_IDs"]
    for customer_id in ids
]

# Get the actual duplicate records
duplicate_records = df[
    df["Customer_ID"].isin(duplicate_customer_ids)
].copy()

# Keep one record from each duplicate pair
golden_records = (
    duplicate_records
    .drop_duplicates(subset=["Email_Normalized"])
    .copy()
)

# Create Master Customer IDs
golden_records["Master_Customer_ID"] = [
    f"MASTER{i:04d}"
    for i in range(1, len(golden_records) + 1)
]

golden_records.head()

,Customer_ID,First_Name,Last_Name,Email,Phone,Date_of_Birth,Country,State,Signup_Date,Customer_Status,Email_Normalized,Phone_Digits,Email_Valid,Email_Quality,DOB_Quality,Signup_Quality,First_Name_Quality,State_Quality,Signup_Year,Master_Customer_ID
76,CUST006573,Arjun,Smith,arjun.smith7573@example.com,141 492 33601,2004-06-09,USA,Illinois,2024-04-22,Inactive,arjun.smith7573@example.com,14149233601,True,Valid,Valid,Valid,Valid,Valid,2024.0,MASTER0001
112,CUST010044,Aarav,Harris,aarav.harris10768@example.com,+1-7865848061,1976-03-12,USA,Washington,2024-04-14,Active,aarav.harris10768@example.com,17865848061,True,Valid,Valid,Valid,Valid,Valid,2024.0,MASTER0002
231,CUST008760,Isabella,Kumar,isabella.kumar9760@example.com,+1-8585642613,2003-11-09,USA,Florida,2020-10-02,Active,isabella.kumar9760@example.com,18585642613,True,Valid,Valid,Valid,Valid,Valid,2020.0,MASTER0003
286,CUST006056,Thomas,Martin,thomas.martin7056@example.com,+49-1624707570,1959-05-11,Germany,Bavaria,2025-04-23,Active,thomas.martin7056@example.com,491624707570,True,Valid,Valid,Valid,Valid,Valid,2025.0,MASTER0004
289,CUST004606,James,Shah,james.shah5606@example.com,NaN,1964-12-28,UK,England,2023-05-14,Active,james.shah5606@example.com,<NA>,True,Valid,Valid,Valid,Valid,Valid,2023.0,MASTER0005


In [107]:
golden_records.shape

(98, 20)

In [108]:
golden_records[
    ["Master_Customer_ID", "Customer_ID", "First_Name",
     "Last_Name", "Email_Normalized"]
].head(10)

,Master_Customer_ID,Customer_ID,First_Name,Last_Name,Email_Normalized
76,MASTER0001,CUST006573,Arjun,Smith,arjun.smith7573@example.com
112,MASTER0002,CUST010044,Aarav,Harris,aarav.harris10768@example.com
231,MASTER0003,CUST008760,Isabella,Kumar,isabella.kumar9760@example.com
286,MASTER0004,CUST006056,Thomas,Martin,thomas.martin7056@example.com
289,MASTER0005,CUST004606,James,Shah,james.shah5606@example.com
316,MASTER0006,CUST008193,Thomas,Khan,thomas.khan9193@example.com
318,MASTER0007,CUST001072,John,Miller,john.miller2072@example.com
372,MASTER0008,CUST001671,Chloe,Roberts,chloe.roberts2671@example.com
426,MASTER0009,CUST009431,Priya,Taylor,priya.taylor10431@example.com
474,MASTER0010,CUST010064,Michael,Brown,michael.brown5479@example.com


In [109]:
golden_records.columns.tolist()

['Customer_ID',
 'First_Name',
 'Last_Name',
 'Email',
 'Phone',
 'Date_of_Birth',
 'Country',
 'State',
 'Signup_Date',
 'Customer_Status',
 'Email_Normalized',
 'Phone_Digits',
 'Email_Valid',
 'Email_Quality',
 'DOB_Quality',
 'Signup_Quality',
 'First_Name_Quality',
 'State_Quality',
 'Signup_Year',
 'Master_Customer_ID']

In [110]:
golden_records[
    [
        "Master_Customer_ID",
        "Customer_ID",
        "First_Name",
        "First_Name_Quality",
        "Email_Normalized",
        "Email_Valid",
        "Email_Quality",
        "Phone_Digits",
        "DOB_Quality",
        "State",
        "State_Quality",
        "Signup_Date",
        "Signup_Quality",
        "Customer_Status"
    ]
].head(10)

,Master_Customer_ID,Customer_ID,First_Name,First_Name_Quality,Email_Normalized,Email_Valid,Email_Quality,Phone_Digits,DOB_Quality,State,State_Quality,Signup_Date,Signup_Quality,Customer_Status
76,MASTER0001,CUST006573,Arjun,Valid,arjun.smith7573@example.com,True,Valid,14149233601,Valid,Illinois,Valid,2024-04-22,Valid,Inactive
112,MASTER0002,CUST010044,Aarav,Valid,aarav.harris10768@example.com,True,Valid,17865848061,Valid,Washington,Valid,2024-04-14,Valid,Active
231,MASTER0003,CUST008760,Isabella,Valid,isabella.kumar9760@example.com,True,Valid,18585642613,Valid,Florida,Valid,2020-10-02,Valid,Active
286,MASTER0004,CUST006056,Thomas,Valid,thomas.martin7056@example.com,True,Valid,491624707570,Valid,Bavaria,Valid,2025-04-23,Valid,Active
289,MASTER0005,CUST004606,James,Valid,james.shah5606@example.com,True,Valid,<NA>,Valid,England,Valid,2023-05-14,Valid,Active
316,MASTER0006,CUST008193,Thomas,Valid,thomas.khan9193@example.com,True,Valid,14847284057,Valid,Texas,Valid,2024-03-09,Valid,Active
318,MASTER0007,CUST001072,John,Valid,john.miller2072@example.com,True,Valid,916403903112,Valid,Karnataka,Valid,2019-10-22,Valid,Suspended
372,MASTER0008,CUST001671,Chloe,Valid,chloe.roberts2671@example.com,True,Valid,918497120447,Valid,Delhi,Valid,2019-01-22,Valid,Active
426,MASTER0009,CUST009431,Priya,Valid,priya.taylor10431@example.com,True,Valid,16106723404,Valid,Texas,Valid,2025-06-21,Valid,Cancelled
474,MASTER0010,CUST010064,Michael,Valid,michael.brown5479@example.com,True,Valid,18127161465,Valid,Washington,Valid,2020-05-31,Valid,Suspended


In [111]:
#Next step: create the final Master Customer table

master_customer_table = (
    golden_records
    .sort_values("Master_Customer_ID")
    .groupby("Master_Customer_ID", as_index=False)
    .first()
)

master_customer_table.shape

(98, 20)

In [112]:
master_customer_table["Master_Customer_ID"].nunique()

98

In [113]:
master_customer_table.isna().sum()

,0
Master_Customer_ID,0
Customer_ID,0
First_Name,0
Last_Name,0
Email,0
Phone,5
Date_of_Birth,4
Country,0
State,2
Signup_Date,1


In [114]:
master_customer_table[
    [
        "Master_Customer_ID",
        "Customer_ID",
        "First_Name",
        "Last_Name",
        "Email_Normalized",
        "Phone_Digits",
        "Date_of_Birth",
        "Country",
        "State",
        "Signup_Date",
        "Customer_Status"
    ]
].head(10)

,Master_Customer_ID,Customer_ID,First_Name,Last_Name,Email_Normalized,Phone_Digits,Date_of_Birth,Country,State,Signup_Date,Customer_Status
0,MASTER0001,CUST006573,Arjun,Smith,arjun.smith7573@example.com,14149233601,2004-06-09,USA,Illinois,2024-04-22,Inactive
1,MASTER0002,CUST010044,Aarav,Harris,aarav.harris10768@example.com,17865848061,1976-03-12,USA,Washington,2024-04-14,Active
2,MASTER0003,CUST008760,Isabella,Kumar,isabella.kumar9760@example.com,18585642613,2003-11-09,USA,Florida,2020-10-02,Active
3,MASTER0004,CUST006056,Thomas,Martin,thomas.martin7056@example.com,491624707570,1959-05-11,Germany,Bavaria,2025-04-23,Active
4,MASTER0005,CUST004606,James,Shah,james.shah5606@example.com,<NA>,1964-12-28,UK,England,2023-05-14,Active
5,MASTER0006,CUST008193,Thomas,Khan,thomas.khan9193@example.com,14847284057,1972-09-30,USA,Texas,2024-03-09,Active
6,MASTER0007,CUST001072,John,Miller,john.miller2072@example.com,916403903112,1979-12-22,India,Karnataka,2019-10-22,Suspended
7,MASTER0008,CUST001671,Chloe,Roberts,chloe.roberts2671@example.com,918497120447,1982-02-27,India,Delhi,2019-01-22,Active
8,MASTER0009,CUST009431,Priya,Taylor,priya.taylor10431@example.com,16106723404,1988-12-01,USA,Texas,2025-06-21,Cancelled
9,MASTER0010,CUST010064,Michael,Brown,michael.brown5479@example.com,18127161465,1962-04-01,USA,Washington,2020-05-31,Suspended


So we're creating two different outputs:

golden_record_map → lineage / duplicate-to-master mapping
master_customer_table → the actual clean golden customer records

That's exactly the distinction we want in a real MDM solution.

In [115]:
final_golden_customer = master_customer_table[
    [
        "Master_Customer_ID",
        "Customer_ID",
        "First_Name",
        "Last_Name",
        "Email_Normalized",
        "Phone_Digits",
        "Date_of_Birth",
        "Country",
        "State",
        "Signup_Date",
        "Customer_Status"
    ]
].copy()

final_golden_customer.shape

(98, 11)

In [127]:
final_golden_customer.sort_values(by=['First_Name','Last_Name']).head(10)

,Master_Customer_ID,Customer_ID,First_Name,Last_Name,Email_Normalized,Phone_Digits,Date_of_Birth,Country,State,Signup_Date,Customer_Status
80,MASTER0081,CUST010063,Aarav,Gupta,aarav.gupta4976@example.com,918534142254,2000-07-24,India,Gujarat,2024-06-16,Active
1,MASTER0002,CUST010044,Aarav,Harris,aarav.harris10768@example.com,17865848061,1976-03-12,USA,Washington,2024-04-14,Active
83,MASTER0084,CUST004740,Aarav,Johnson,aarav.johnson5740@example.com,6587866323,1991-07-12,Singapore,Central Region,2020-12-11,Active
32,MASTER0033,CUST010042,Aarav,Kumar,aarav.kumar4269@example.com,61490490785,1972-02-05,Australia,Queensland,2025-06-14,Active
44,MASTER0045,CUST010035,Aarav,Thomas,aarav.thomas4010@example.com,491517916477,1995-06-10,Germany,Hamburg,2024-12-01,Suspended
20,MASTER0021,CUST004550,Aisha,Anderson,aisha.anderson5550@example.com,447122860906,1988-02-15,UK,England,2022-11-30,Cancelled
60,MASTER0061,CUST010014,Aisha,Williams,aisha.williams3017@example.com,15356112916,1994-04-20,USA,Washington,2021-03-21,Active
37,MASTER0038,CUST007016,Alex,Anderson,alex.anderson8016@example.com,918723437425,1959-01-29,India,None,2021-05-11,Active
93,MASTER0094,CUST004734,Alex,Chen,alex.chen5734@example.com,919178184642,1967-07-27,India,Telangana,2024-11-06,Cancelled
81,MASTER0082,CUST010081,Alex,Kumar,alex.kumar3695@example.com,17482068523,1962-06-27,Canada,Ontario,2021-07-26,Active


In [117]:
# 1. Every source Customer_ID should map to exactly one Master_Customer_ID
golden_records["Customer_ID"].nunique(), golden_records["Master_Customer_ID"].nunique()

(98, 98)

In [118]:
# 2. Check whether any Customer_ID appears more than once
golden_records["Customer_ID"].duplicated().sum()

np.int64(0)

In [119]:
# 3. Check whether any Master_Customer_ID appears more than once
master_customer_table["Master_Customer_ID"].duplicated().sum()

np.int64(0)

1. golden_customer_table.csv
The 98 mastered customer records.
2. customer_master_mapping.csv
The mapping from the original Customer_IDs to Master_Customer_IDs. This is particularly important from an MDM perspective because it shows how duplicate source records were consolidated.
3. data_quality_summary.csv
Your issue/treatment table:
80 exact duplicate rows removed
98 duplicate customer groups identified
104 invalid emails flagged
25 future DOBs handled
21 future signup dates handled
etc.

Then we'll make one final README/project summary explaining the business problem and methodology.

In [120]:
golden_records.to_csv(
    folder_path + "golden_customer_table.csv",
    index=False
)

In [121]:
customer_master_mapping = golden_records[
    ["Customer_ID", "Master_Customer_ID"]
].copy()

customer_master_mapping.to_csv(
    folder_path + "customer_master_mapping.csv",
    index=False
)

In [122]:
data_quality_summary = pd.DataFrame({
    "Issue": [
        "Exact duplicate rows",
        "Duplicate Customer_ID",
        "Missing Phone",
        "Missing Email",
        "Invalid Email",
        "Future DOB",
        "Missing DOB",
        "Missing State",
        "Missing First Name",
        "Future Signup Date",
        "DOB >= Signup Date"
    ],

    "Records": [
        80,
        0,
        255,
        118,
        104,
        25,
        154,
        102,
        80,
        21,
        0
    ],

    "Treatment": [
        "Removed",
        "No action",
        "Retained",
        "Retained",
        "Flagged",
        "Set to missing",
        "Retained",
        "Retained",
        "Retained",
        "Set to missing",
        "No action"
    ],

    "Business_Rationale": [
        "Redundant records",
        "Customer_ID is unique",
        "No reliable value available",
        "No reliable value available",
        "Do not fabricate/correct without source",
        "Invalid future date",
        "Missing allowed; no reliable replacement",
        "Cannot reliably infer state",
        "Cannot reliably infer from email",
        "Invalid future date",
        "No chronological violations"
    ]
})

data_quality_summary.to_csv(
    folder_path + "data_quality_summary.csv",
    index=False
)

## SUMMARY

In [130]:
#Cell 1: Data Quality Summary
data_quality_summary = pd.DataFrame({
    "Data Quality Check": [
        "Records analyzed",
        "Missing Email",
        "Invalid Email",
        "Missing Phone",
        "Missing State",
        "Missing First Name",
        "Future DOB",
        "Future Signup Date"
    ],
    "Result": [
        len(df),
        df["Email"].isna().sum(),
        (df["Email_Quality"] == "Invalid").sum(),
        df["Phone_Digits"].isna().sum(),
        df["State"].isna().sum(),
        df["First_Name"].isna().sum(),
        (df["DOB_Quality"] == "Invalid_Future").sum(),
        (df["Signup_Quality"] == "Invalid_Future").sum()
    ]
})
data_quality_summary.style.hide(axis="index")

Data Quality Check,Result
Records analyzed,10100
Missing Email,118
Invalid Email,122
Missing Phone,255
Missing State,102
Missing First Name,80
Future DOB,25
Future Signup Date,21


In [131]:
duplicate_summary = pd.DataFrame({
    "Duplicate Detection Metric": [
        "Duplicate customer groups",
        "Source records involved",
        "Records per duplicate group"
    ],
    "Result": [
        len(pair_analysis),
        len(duplicate_records),
        "2"
    ]
})

duplicate_summary.style.hide(axis="index")

Duplicate Detection Metric,Result
Duplicate customer groups,98
Source records involved,196
Records per duplicate group,2


In [132]:
match_validation_summary = pd.DataFrame({
    "Match Validation": [
        "Duplicate groups evaluated",
        "Groups with exact attribute agreement",
        "Groups requiring attribute review"
    ],
    "Result": [
        len(pair_analysis),
        pair_analysis["Exact_Match"].sum(),
        (~pair_analysis["Exact_Match"]).sum()
    ]
})

match_validation_summary.style.hide(axis="index")

Match Validation,Result
Duplicate groups evaluated,98
Groups with exact attribute agreement,98
Groups requiring attribute review,0


In [134]:
golden_showcase = final_golden_customer[
    [
        "Master_Customer_ID",
        "Customer_ID",
        "First_Name",
        "Last_Name",
        "Email_Normalized",
        "Phone_Digits",
        "Country",
        "Customer_Status"
    ]
].head(10)

golden_showcase.style.hide(axis="index")

Master_Customer_ID,Customer_ID,First_Name,Last_Name,Email_Normalized,Phone_Digits,Country,Customer_Status
MASTER0001,CUST006573,Arjun,Smith,arjun.smith7573@example.com,14149233601,USA,Inactive
MASTER0002,CUST010044,Aarav,Harris,aarav.harris10768@example.com,17865848061,USA,Active
MASTER0003,CUST008760,Isabella,Kumar,isabella.kumar9760@example.com,18585642613,USA,Active
MASTER0004,CUST006056,Thomas,Martin,thomas.martin7056@example.com,491624707570,Germany,Active
MASTER0005,CUST004606,James,Shah,james.shah5606@example.com,,UK,Active
MASTER0006,CUST008193,Thomas,Khan,thomas.khan9193@example.com,14847284057,USA,Active
MASTER0007,CUST001072,John,Miller,john.miller2072@example.com,916403903112,India,Suspended
MASTER0008,CUST001671,Chloe,Roberts,chloe.roberts2671@example.com,918497120447,India,Active
MASTER0009,CUST009431,Priya,Taylor,priya.taylor10431@example.com,16106723404,USA,Cancelled
MASTER0010,CUST010064,Michael,Brown,michael.brown5479@example.com,18127161465,USA,Suspended


In [135]:
final_validation_summary = pd.DataFrame({
    "Final Validation Check": [
        "Golden records created",
        "Unique Master Customer IDs",
        "Duplicate Master Customer IDs",
        "Source-to-master mapping records"
    ],
    "Result": [
        len(golden_records),
        golden_records["Master_Customer_ID"].nunique(),
        golden_records["Master_Customer_ID"].duplicated().sum(),
        len(golden_record_map)
    ]
})

final_validation_summary.style.hide(axis="index")

Final Validation Check,Result
Golden records created,98
Unique Master Customer IDs,98
Duplicate Master Customer IDs,0
Source-to-master mapping records,98
